controllers/disk.go

```diff
@@ -42,118 +42,251 @@ func DiskCreate(c *gin.Context) {
 		return
 	}
 
-	userUUID := _userUUID
-
-	disk := DriveFactory.NewDisk(provider.Type)
+	volume := models.Transport.GetVolume(userUUID, volumeUUID)
+	_disk := dbo.Disk{
+		AbstractDatabaseObject: dbo.AbstractDatabaseObject{
+			UUID: uuid.New(),
+		},
+		UserUUID:     userUUID,
+		VolumeUUID:   volumeUUID,
+		ProviderUUID: provider.UUID,
+		Credentials:  requestBody.Credentials.ToString(),
+		Provider:     *provider,
+	}
+	disk := models.CreateDisk(models.CreateDiskMetadata{
+		Disk:   &_disk,
+		Volume: volume,
+	})
 	if disk == nil {
 		c.JSON(422, responses.NewValidationErrorResponseSingle(constants.VAL_PROVIDER_NOT_SUPPORTED, "ProviderUUID", "Provided ProviderUUID is not a supported provider"))
 		return
 	}
+	db.DB.DatabaseHandle.Omit("Provider", "Volume").Create(_disk)
 
-	disk.SetUUID(uuid.New())
-
-	// Get volume handle
-	volume := models.Transport.GetVolume(userUUID, volumeUUID)
-	volume.AddDisk(disk.GetUUID(), disk)
-	providerUUID, _ := uuid.Parse(requestBody.ProviderUUID)
-
-	if provider.Type == constants.PROVIDER_TYPE_SFTP || provider.Type == constants.PROVIDER_TYPE_FTP {
-		disk.CreateCredentials(requestBody.Credentials)
-		db.DB.DatabaseHandle.Create(disk.GetDiskDBO(userUUID, providerUUID, volumeUUID))
+	_, ok := disk.(OAuthDisk.OAuthDisk)
+	if ok {
+		config := disk.(OAuthDisk.OAuthDisk).GetConfig()
+		authCode = config.AuthCodeURL("state-token", oauth2.AccessTypeOffline)
 	}
 
-	if provider.Type == constants.PROVIDER_TYPE_ONEDRIVE || provider.Type == constants.PROVIDER_TYPE_GDRIVE {
-		db.DB.DatabaseHandle.Create(disk.GetDiskDBO(userUUID, providerUUID, volumeUUID))
-		config := disk.(disk2.OAuthDisk).GetConfig()
-		authCode = config.AuthCodeURL("state-token", oauth2.AccessTypeOffline)
+	// load full database object with a provider and a volume to return
+	err = db.DB.DatabaseHandle.Where("uuid = ?", disk.GetUUID().String()).Preload("Provider").Preload("Volume").Find(&_disk).Error
+	if err != nil {
+		// this should never happen
+		c.JSON(422, responses.NewValidationErrorResponse(err, "Could not validate database change"))
```

Comment:
(*) To chyba nie zadziała - NewValidationErrorResponse zwraca response dla Errora, który jest powiązany z walidacją pól jsonowych - obiekt Error ma wtedy specyficzną strukturę na podstawie której tworzony jest response zawierający listę pól i związanych z nimi błędów. Wydaje mi się, że jeśli chcielibyśmy zwrócić 422 to lepszym wyborem byłby "konstruktor" NewValidationErrorResponseSingle, ale może w tym przypadku zwrócić 500 jako błąd wewnętrzny (tzn. nie znaleziono dysku w bazie)?